src/test/versionbits_tests.cpp

```diff
@@ -242,81 +275,121 @@ BOOST_AUTO_TEST_CASE(versionbits_test)
     }
 }
 
-BOOST_AUTO_TEST_CASE(versionbits_computeblockversion)
+static void check_computeblockversion(const Consensus::Params& params, Consensus::DeploymentPos dep)
 {
+    // This implicitly uses versionbitscache, so clear it every time
+    versionbitscache.Clear();
+
     // Check that ComputeBlockVersion will set the appropriate bit correctly
-    // on mainnet.
-    const auto chainParams = CreateChainParams(*m_node.args, CBaseChainParams::MAIN);
-    const Consensus::Params &mainnetParams = chainParams->GetConsensus();
 
-    // Use the TESTDUMMY deployment for testing purposes.
-    int64_t bit = mainnetParams.vDeployments[Consensus::DEPLOYMENT_TESTDUMMY].bit;
-    int64_t nStartTime = mainnetParams.vDeployments[Consensus::DEPLOYMENT_TESTDUMMY].nStartTime;
-    int64_t nTimeout = mainnetParams.vDeployments[Consensus::DEPLOYMENT_TESTDUMMY].nTimeout;
+    int64_t bit = params.vDeployments[dep].bit;
+    int64_t nStartTime = params.vDeployments[dep].nStartTime;
+    int64_t nTimeout = params.vDeployments[dep].nTimeout;
+    int min_activation_height = params.vDeployments[dep].min_activation_height;
 
-    assert(nStartTime < nTimeout);
+    // should not be any signalling for first block
+    BOOST_CHECK_EQUAL(ComputeBlockVersion(nullptr, params), VERSIONBITS_TOP_BITS);
+
+    // always/never active deployments shouldn't need to be tested further
+    if (nStartTime == Consensus::BIP9Deployment::ALWAYS_ACTIVE) return;
+    if (nStartTime == Consensus::BIP9Deployment::NEVER_ACTIVE) return;
+
+    BOOST_REQUIRE(nStartTime < nTimeout);
+    BOOST_REQUIRE(nStartTime >= 0);
+    BOOST_REQUIRE(nTimeout <= std::numeric_limits<uint32_t>::max() || nTimeout == Consensus::BIP9Deployment::NO_TIMEOUT);
+    BOOST_REQUIRE(0 <= bit && bit < 32);
```

Comment:
It's already checked via `BOOST_REQUIRE(((1 << bit) & VERSIONBITS_TOP_MASK) == 0);` -- the "requires" are there because the later checks will break if these aren't true; if it were possible for `bit` to be  `NUM_BITS` or greater, but not clash with `TOP_MASK` the other tests would be fine.

(I had also thought `versionbits_sanity` was checking the bit was in `[0, NUM_BITS)` so didn't "duplicate" the test here, but it's not actually doing that -- perhaps I saw that in some patch implementing BIP320)

Reply:
Noting that this check is required before computing on the bit (https://en.cppreference.com/w/cpp/language/operator_arithmetic):

> In any case, if the value of the right operand is negative or is greater or equal to the number of bits in the promoted left operand, the behavior is undefined. 

This also impacts where, if taken, https://github.com/bitcoin/bitcoin/pull/21377/commits/754ab11f2085a0c5e417cfbba80cd024fc8ea67a#r609525131 should be created.